### ISSUE TITLE
Generic gulp-task

### ISSUE THREAD

**Issue by graphefruit** [QuantumBytesInc/RelicsOfAnnorath#55]
Actual solution:
We've to run the gulp task ourself on client side and commit these changes.

Problem:
When someone changes files and we want to release it, we need to run the task ourself (gulp needs to be installed etc.) and then commit these changes.

Also we need to include all JavaScript/CSS - Files by our self.
Therefore we run into the issue if new files are created they are not automatically added into the gulp-task.

We need to provide a more automatic way here.

Solution:
We need to implement this task when the docker-container is set up.
1. Pull git
2. Run gulp task
3. Collect static 
4. Push to amazon
5. Finish.

@ManuelGysin I need your help here

**Comment by graphefruit** [comment]
@ManuelGysin Could our Jenkins server maybe help us here? Or do we need to do this in the Docker-Files?

**Comment by graphefruit** [comment]
Still hold actualy, because a not prior task

**Comment by ManuelGysin** [comment]
Please provide needed steps, we will integrate this within drone.

**Comment by graphefruit** [comment]
I'll provide steps tomorrow

**Comment by graphefruit** [comment]
Sorry for the delay.
Here we go:

1. Pull git (Get latest JS, CSS)
2. Switch to needed directory and run gulp task
2.1 Gulp task will grab all JavaScript files and concat them to one big file, which will be minified.
2.2 Gulp task will grab all CSS files and concat them to one big file, which will be minified
2.2.1 This task needs to handle the SASS-Compile aswell (I already saw one compile directive with gulp for this, but need to grab into it)
2.3. Gulp task will save this JavaScript and CSS file into an distribution folder.
2.3.1 We need to take care about the CSS-File and the relative image handling - maybe we should add a variable when building the SCSS which inherits the full path to avoid conflicts?
3. Collect static + Amazon Push
(4. Maybe add these two files aswell into the git to test them on stage system?)
Task is finished
A sample gulpfile is visible here: https://github.com/QuantumBytesInc/ROAP/blob/master/project/gui/static/gulpfile.js

Requirements:
```
var gulp = require('gulp');
var concat = require('gulp-concat');
var uglify = require('gulp-uglify');
var minifyHtml = require('gulp-minify-html');
var ngTemplate = require('gulp-ng-template');
var cleanCSS = require('gulp-clean-css');
```
This gulp file needs to be changed to work synchron and wait for one task to finish, before the next one will be triggered.
This actual gulp task can't do this, I'll provide a new one until next week.

The requirements will get bigger when changing to synchron task and adding sass support.

@ManuelGysin You need a further meeting to discuss this?

**Comment by graphefruit** [comment]
@ManuelGysin One more comment: For GULP we need Nodejs and then the gulp-packages installed in the container - you think this is without bigger problems possible?

**Comment by ManuelGysin** [comment]
I just need the requirements and the script.

**Comment by graphefruit** [comment]
I'll transform the ROAG-script and then we can start playing with it 👍

**Comment by ManuelGysin** [comment]
As FYI:

- Drone launches a specific docker container, this container is not related to our normal containers for web applications
- Drone executes your script
- Drone may commit the result and sync to s3
- aws lambda will trigger cloudfront sync
- we are good to go

**Comment by graphefruit** [comment]
Actual requirements:

- [ ] - NodeJS

NodeJS Packages:
- [ ] - npm install --save-dev gulp
- [ ] - npm install --save-dev gulp-sequence
- [ ] - npm install --save-dev gulp-sass
- [ ] - npm install --save-dev gulp-concat
- [ ] - npm install --save-dev gulp-uglify
- [ ] - npm install --save-dev gulp-minify-html
- [ ] - npm install --save-dev gulp-ng-template
- [ ] - npm install --save-dev yargs
- [ ] - npm install --save-dev gulp-clean
- [ ] - npm install --save-dev gulp-insert
- [ ] - npm install --save-dev gulp-clean-css
- [ ] - npm install --save-dev gulp-replace
- [ ] - npm install --save-dev require-dir

I don't know if you want to install them globaly or `save-dev`, just use what fits our need

The task can be started like:
`gulp --URL http://our-url.de`

The provided url will be taken and replace all placeholders in `.SCSS` files with placeholder: `{{URL}}`

The javascript-template urls will be generated like:
`prefix: argv.URL + '/data/static/gui/templates/'`

So our url needs to inherit `/stage/` or `/production/`

Needs to be discussed:
Will we provide a version flag aswell? Then I can add them into the copyright information

Needs to be done on my side:
- [ ] Change CSS to SCSS folder
- [ ] Remove all `.css` and `.map` files
- [ ] Adjust the image urls with placeholder inside `.scss`
- [ ] Adjust the needed JavaScript-Files
- [ ] Run the script localy and test
- [ ] Finish and commit

**Comment by graphefruit** [comment]
We also need to take care about ROAG and ROAP, I think we should start with ROAG because its more important.
ROAP doesn't has any `.SCSS` files, but the rest should nearly stay the same.

**Comment by graphefruit** [comment]
Extend the gulp file:

Gulp file needs to be extended to grab the `--BRANCH development` parameter.
If the BRANCH is not `master` we'll copy all `js` files etc.
If the BRANCH is `master` we'll just copy the minified `app.min.js`

Now we need to extend the logic to copy the files from:
`/source/` to `/static/gui/` with all needed references for production and development system.
From there the files will be copied to amazon with `collectstatic`

**Comment by graphefruit** [comment]
@ManuelGysin We need three more modules:
```
npm install gulp-copy -g
npm install merge-stream -g
npm install gulp-rename -g
```
Complete command:
`npm install gulp-copy merge-stream gulp-rename -g`

Other scripts are checked in for directly testing.

**Comment by graphefruit** [comment]
Current status:
The script copies the needed files / folders for production OR development.
This depends on the passed `--BRANCH` parameter.

If `--BRANCH` is `master` it will run production, else it will always run development.
All files has been outsourced to: 
`/gui/source/...`

All files will be copied to:
`/gui/static/gui/...`

From there the collectstatic will grab them and copy all to amazon.

Actual gulp-parameters:
- --BRANCH - Actual Branch
- --URL - The URL to Amazon for production / pevelopment
- --VERSION - The version / hash for the copyright information

All parameters are required.

**Comment by graphefruit** [comment]
Task update:
JavaScript and CSS-Files needs to be inserted into the gulp building process.

JS:
Dependency-Injection needs to be tested after minifying (This is already working in ROAP)

CSS: 
Url's need to be adjusted with placeholders.

The JavaScript-Files needs to be added in a special order, because we need e.g. to add angular and the other frameworks before we can add our own logic.

@ManuelGysin 
We need to adjust ROAP aswell, I'd suggest to do this after successfully changing ROAG.
Shall we create a new issue for it?

**Comment by ManuelGysin** [comment]
Done both tasks.

**Comment by graphefruit** [comment]
Gulp tasks for ROAG and ROAP are finished now on development side.

Adding `show_live=true` as URL-Parameter will render the page as minified and live page.

Drone task needs a bit fine tuning for master releases now and then we're good to go.

**Comment by graphefruit** [comment]
After discussion: The `sqlite/exports.sqlite` will be generated from python and needs to be copied directly to `/static/gui/sqlite/exports.sqlite` @ManuelGysin 

Second the translations file which will be also generated from python needs to be copied into `/source/js/static/translations.js` from here the gulptask need to grab this file for minification

**Comment by graphefruit** [comment]
Also we've changed the interface includings for caller, actions, codes on ROAG to the roa-constants.

**Comment by graphefruit** [comment]
Gulp task done for ROAG / ROAP.
We'll need to fix them for master commits, but this will be a following task.
